indra/newview/llinventorylistener.cpp

```diff
@@ -0,0 +1,276 @@
+{
+    Response response(LLSD(), data);
+    LLUUID folder_id(data["folder_id"].asUUID());
+    LLViewerInventoryCategory *cat = gInventory.getCategory(folder_id);
+    if (!cat)
+    {
+        return response.error(stringize("Folder ", std::quoted(data["folder_id"].asString()), " was not found"));
+    }
+    LLInventoryModel::cat_array_t  cat_array;
+    LLInventoryModel::item_array_t item_array;
+
+    LLFilteredCollector collector = LLFilteredCollector(data);
+
+    gInventory.collectDescendentsIf(folder_id, cat_array, item_array, LLInventoryModel::EXCLUDE_TRASH, collector);
+
+    for (LLInventoryModel::item_array_t::iterator iter = item_array.begin(); iter != item_array.end(); iter++)
+    {
+        add_item_info(response, *iter);
+    }
+    for (LLInventoryModel::cat_array_t::iterator iter = cat_array.begin(); iter != cat_array.end(); iter++)
+    {
+        add_cat_info(response, *iter);
+    }
+}
+
+LLFilteredCollector::LLFilteredCollector(LLSD const &data) :
+    mType(LLAssetType::EType::AT_UNKNOWN),
+    mLinkFilter(INCLUDE_LINKS),
+    mItemLimit(MAX_ITEM_LIMIT),
+    mItemCount(0)
+{
+    if (data.has("name"))
+    {
+        mName = data["name"].asString();
+    }
+    if (data.has("desc"))
+    {
+        mDesc = data["desc"].asString();
+    }
+    if (data.has("type"))
+    {
+        mType = LLAssetType::lookup(data["type"]);
+    }
+    if (data.has("filter_links"))
+    {
+        if (data["filter_links"] == "EXCLUDE_LINKS")
+        {
+            mLinkFilter = EXCLUDE_LINKS;
+        }
+        else if (data["filter_links"] == "ONLY_LINKS")
+        {
+            mLinkFilter = ONLY_LINKS;
+        }
+    }
+    if (data.has("item_limit"))
+    {
+        mItemLimit = llclamp(data["item_limit"].asInteger(), 1, MAX_ITEM_LIMIT);
+    }
+}
+
+bool LLFilteredCollector::operator()(LLInventoryCategory *cat, LLInventoryItem *item)
+{
+    bool passed = checkagainstType(cat, item);
+    passed = passed && checkagainstNameDesc(cat, item);
+    passed = passed && checkagainstLinks(cat, item);
+
+    if (passed)
+    {
+        ++mItemCount;
+    }
+    return passed;
+}
+
+bool LLFilteredCollector::exceedsLimit()
+{
+    return (mItemLimit <= mItemCount);
+}
+
+bool LLFilteredCollector::checkagainstNameDesc(LLInventoryCategory *cat, LLInventoryItem *item)
+{
+    std::string name, desc;
+    bool passed(true);
+    if (cat)
+    {
+        if (!mDesc.empty()) return false;
+        name = cat->getName();
+    }
+    if (item)
+    {
+        name = item->getName();
+        passed = (mDesc.size() ? item->getDescription().find(mDesc) != std::string::npos : true);
+    }
+
+    return passed && (mName.size() ? name.find(mName) != std::string::npos : true);
+}
+
+bool LLFilteredCollector::checkagainstType(LLInventoryCategory *cat, LLInventoryItem *item)
+{
+    if (mType == LLAssetType::AT_UNKNOWN)
+    {
+        return true;
+    }
+    if (mType == LLAssetType::AT_CATEGORY)
```

Comment:
Seems like lines 254-260 could pretty readily be consolidated into something more like lines 261-264.